kernel/aster-nix/src/fs/utils/page_cache.rs

```diff
@@ -85,16 +85,185 @@ impl Debug for PageCache {
     }
 }
 
+struct ReadaheadState {
+    /// Current readahead window start.
+    ra_start: usize,
+    /// Current readahead window size.
+    ra_size: usize,
+    /// Look ahead position in current window.
+    lookahead_index: usize,
+    /// Last page visited.
+    prev_page: usize,
+    /// Maximum window size.
+    ra_pages: usize,
+    /// Readahead requests waiter.
+    waiter: BioWaiter,
+}
+
+impl ReadaheadState {
+    const INVALID_PAGE_IDX: usize = 0xFFFFFFFFFFFFFFFFusize;
+    const INIT_RA_PAGES: usize = 4;
+    const MAX_RA_PAGES: usize = 32;
+
+    pub fn new() -> Self {
+        Self {
+            ra_start: Self::INVALID_PAGE_IDX,
+            ra_size: 0,
+            lookahead_index: Self::INVALID_PAGE_IDX,
+            prev_page: Self::INVALID_PAGE_IDX,
+            ra_pages: Self::MAX_RA_PAGES,
+            waiter: BioWaiter::new(),
+        }
+    }
+
+    pub fn set_ra_page(&mut self, page: usize) {
+        self.ra_pages = if page > Self::MAX_RA_PAGES {
+            Self::MAX_RA_PAGES
+        } else if page < Self::INIT_RA_PAGES {
+            Self::INIT_RA_PAGES
+        } else {
+            page
+        };
+    }
+
+    fn init_ra_range(&mut self, idx: usize, npages: usize) -> bool {
+        // Set initial readahead window.
+        self.ra_start = idx;
+        if self.ra_start >= npages {
+            self.ra_start = Self::INVALID_PAGE_IDX;
+            return false;
+        }
+        // Set initial window size. Make sure no overflow.
+        self.ra_size = Self::INIT_RA_PAGES.min(npages - self.ra_start);
+        // Fully pipelined.
+        self.lookahead_index = self.ra_start;
+        true
+    }
+
+    fn update_ra_range(&mut self, npages: usize) -> bool {
+        // Push forward window.
+        self.ra_start += self.ra_size;
+        if self.ra_start >= npages {
+            self.ra_start = Self::INVALID_PAGE_IDX;
+            return false;
+        }
+        // Calculate new window size.
+        self.ra_size = (self.ra_size * 2).min(self.ra_pages);
+        // Make sure no overflow.
+        self.ra_size = self.ra_size.min(npages - self.ra_start);
+        // Fully pipelined.
+        self.lookahead_index = self.ra_start;
+        true
+    }
+
+    fn is_sequential(&self, idx: usize) -> bool {
+        self.prev_page != Self::INVALID_PAGE_IDX
+            && (idx == self.prev_page || idx == self.prev_page + 1)
+    }
+
+    fn readahead_index(&self) -> usize {
+        self.ra_start + self.ra_size
+    }
+
+    fn get_ra_range(&self) -> Option<Range<usize>> {
+        if self.ra_start == Self::INVALID_PAGE_IDX {
+            return None;
+        }
+        Some(self.ra_start..self.ra_start + self.ra_size)
+    }
+
+    fn is_all_req_completed(&self) -> bool {
```

Comment:
We can use `has_completed_all_requests`.

Reply:
Fixed.